File: android/mathrenderer/src/main/java/io/github/dongyuzhao/mathrenderer/MathRenderer.kt
package io.github.dongyuzhao.mathrenderer

import android.content.Context
import android.graphics.Color

/**
 * MathRenderer — pure MathJax-driven LaTeX formula renderer for Android.
 *
 * Runs the MathJax bundle inside an isolated JavaScript engine to produce SVG,
 * then rasterises the SVG to a [FormulaImage] using an in-house Android
 * Canvas/Path SVG subset renderer.
 *
 * ## Quick start
 *
 * ```kotlin
 * val image = MathRenderer.render(context, "x^2 + y^2 = r^2")
 * // image?.bitmap → Bitmap ready to display
 * // image?.baselineOffsetDp → for inline text alignment
 * ```
 */
object MathRenderer {

    /**
     * Renders a LaTeX formula to a [FormulaImage].
     *
     * Must be called from a coroutine (suspend function). The JS evaluation
     * happens off the main thread automatically.
     *
     * @param context    Android context (Application context is recommended).
     * @param latex      LaTeX source without surrounding `$` delimiters.
     * @param options    Rendering options (mode, font size, scale).
     * @param color      Fill colour (ARGB int); defaults to black.
     * @param density    Screen density for rasterisation (from `DisplayMetrics.density`).
     */
    suspend fun render(
        context: Context,
        latex: String,
        options: FormulaRenderOptions = FormulaRenderOptions(),
        color: Int = Color.BLACK,
        density: Float = context.resources.displayMetrics.density,
    ): FormulaImage? {
        val svg = MathJaxBridge.shared(context).svg(latex, options) ?: return null
        return FormulaRasterizer(density).image(
            svg = svg,
            color = color,
            scale = density * options.scale.toFloat(),
            fontSizeDp = options.fontSize.toFloat(),
        )
    }

    /**
     * Returns only the [MathJaxSVG] for a LaTeX formula (no rasterisation).
     */
    suspend fun svg(
        context: Context,
        latex: String,
        options: FormulaRenderOptions = FormulaRenderOptions(),
    ): MathJaxSVG? = MathJaxBridge.shared(context).svg(latex, options)
}

File: swift/Sources/MathRenderer/MathRenderer.swift
/// MathRenderer — pure MathJax-driven LaTeX formula renderer.
///
/// Runs MathJax inside JavaScriptCore to produce SVG, then rasterises the SVG
/// to a platform image (UIImage / NSImage) using an in-house CoreGraphics
/// SVG subset renderer.
///
/// ## Quick start
///
/// ```swift
/// // Inline formula, system scale, 16pt font, black
/// if let image = await MathRenderer.render("x^2 + y^2 = r^2") {
///     // image.image    → UIImage / NSImage
///     // image.size     → logical size in points
///     // image.baselineOffset → for inline text placement
/// }
///
/// // Display (block) mode, custom colour
/// let opts = FormulaRenderOptions(standalone: true, fontSize: 20, scale: 2)
/// if let image = await MathRenderer.render("\\int_0^\\infty e^{-x^2}\\,dx", options: opts) { ... }
/// ```
public enum MathRenderer {

#if canImport(CoreGraphics)
    /// Renders a LaTeX formula string to a platform image.
    ///
    /// - Parameters:
    ///   - latex:   LaTeX source without surrounding `$` delimiters.
    ///   - options: Rendering options (mode, font size, scale).
    ///   - color:   Fill colour; `nil` defaults to black.
    ///   - scale:   Screen scale; uses `options.scale` when this parameter is `nil`.
    /// - Returns:   A `FormulaImage` on success, or `nil` if rendering failed.
    @MainActor
    public static func render(
        _ latex: String,
        options: FormulaRenderOptions = FormulaRenderOptions(),
        color: CGColor? = nil,
        scale: CGFloat? = nil
    ) async -> FormulaImage? {
        guard let svg = await MathJaxBridge.shared.svg(for: latex, options: options) else {
            return nil
        }
        let resolvedScale = scale ?? CGFloat(options.scale)
        return FormulaRasterizer.shared.image(
            from: svg,
            color: color,
            scale: resolvedScale,
            fontSize: CGFloat(options.fontSize)
        )
    }

    /// Returns only the SVG markup for a LaTeX formula (no rasterisation).
    public static func svg(
        _ latex: String,
        options: FormulaRenderOptions = FormulaRenderOptions()
    ) async -> MathJaxSVG? {
        await MathJaxBridge.shared.svg(for: latex, options: options)
    }
#endif
}
